src/shims/unix/fs.rs

```diff
@@ -1659,6 +1666,138 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+            let parts = unsafe { win_encoded.align_to::<u8>() };
+            parts.1
+        } else {
+            TEMPFILE_TEMPLATE_STR.as_bytes()
+        };
+
+        // At this point we have one `&[u8]` that represents the template and one `&[u8]`
+        // that represents the suffix, both with the same (_target_) encoding.
+        // Critically, the encoding may differ from the host / where _this_ code is running!
+
+        // Now we figure out the index of the slice we expect to contain the suffix.
+        let start_pos = template_bytes.len().saturating_sub(suffix_bytes.len());
+        let end_pos = template_bytes.len();
+        let last_six_char_bytes = &template_bytes[start_pos..end_pos];
+
+        // If we don't find the suffix, it is an error.
+        if last_six_char_bytes != suffix_bytes {
+            let einval = this.eval_libc("EINVAL")?;
+            this.set_last_error(einval)?;
+            return Ok(-1);
+        }
+
+        // At this point we know we have 6 ASCII 'X' characters as a suffix regardless of
+        // host and target encoding.
+
+        // From <https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/sysdeps/posix/tempname.c#L175>
+        let substitutions = &[
+            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q',
+            'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
+            'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y',
+            'Z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+        ];
+        // Create the random distribution of substitutions.
+        let between = Uniform::from(0..substitutions.len());
+        let mut rng = this.machine.rng.clone();
+
+        // The file is opened with specific options, the implementation of which depends on
+        // the host OS.
+        let mut fopts = OpenOptions::new();
+        fopts.read(true).write(true).create_new(true);
+
+        #[cfg(target_family = "unix")]
+        {
+            use std::os::unix::fs::OpenOptionsExt;
+            fopts.mode(0o600);
+            // Do not allow others to read or modify this file.
+            fopts.custom_flags(libc::O_EXCL);
+        }
+        #[cfg(target_family = "windows")]
+        {
+            use std::os::windows::fs::OpenOptionsExt;
+            // Do not allow others to read or modify this file.
+            fopts.share_mode(0);
+        }
+
+        // If the generated file already exists, we will try again this many times.
+        let attempts = if this.eval_context_ref().tcx.sess.target.os == "linux" {
+            TEMPFILE_ATTEMPTS_GLIBC
+        } else {
+            TEMPFILE_ATTEMPTS_MIN
+        };
+
+        for _ in 0..attempts {
+            let unique_suffix = between
+                .sample_iter(rng.get_mut())
+                .take(6)
+                .map(|idx| substitutions[idx])
```

Comment:
You can avoid this indirection through indexing by using https://docs.rs/rand/0.6.5/rand/seq/trait.SliceRandom.html#tymethod.choose

Reply:
TIL.